worker/changestream/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,326 @@
+func (q *EventQueue) Wait() error {
+	return q.catacomb.Wait()
+}
+
+func (q *EventQueue) unsubscribe(subscriptionID uint64) {
+	go func() {
+		select {
+		case <-q.catacomb.Dying():
+			return
+		case q.unsubscriptionCh <- subscriptionID:
+		}
+	}()
+}
+
+func (q *EventQueue) loop() error {
+	defer func() {
+		q.subscriptions = nil
+		q.subscriptionsByNS = nil
+	}()
+
+	for {
+		select {
+		case <-q.catacomb.Dying():
+			return q.catacomb.ErrDying()
+
+		case ch, ok := <-q.stream.Changes():
+			// If the stream is closed, we expect that a new worker will come
+			// again using the change stream worker infrastructure. In this case
+			// just ignore and close out.
+			if !ok {
+				q.logger.Infof("change stream change channel is closed")
+				return nil
+			}
+
+			subs := q.gatherSubscriptions(ch)
+			for _, sub := range subs {
+				// Ensure we check tomb dying as the handling logic is
+				// synchronous and blocking. This is to effectively handle
+				// back pressure on the stream.
+				select {
+				case <-q.catacomb.Dying():
+					return q.catacomb.ErrDying()
+				default:
+				}
+				sub.handler(ch)
+			}
+
+		case subOpt := <-q.subscriptionCh:
+			sub := subOpt.subscription
+
+			// Create a new subscription and assign a unique ID to it.
+			q.subscriptions[sub.id] = sub
+
+			// No options were supplied, just add it to the all bucket, so
+			// they'll be included in every dispatch.
+			if len(subOpt.opts) == 0 {
+				q.subscriptionsAll[sub.id] = struct{}{}
+				continue
+			}
+
+			// Register filters to route changes matching the subscription criteria to
+			// the newly crated subscription.
+			for _, opt := range subOpt.opts {
+				namespace := opt.Namespace()
+				q.subscriptionsByNS[namespace] = append(q.subscriptionsByNS[namespace], &eventFilter{
+					subscriptionID: sub.id,
+					changeMask:     opt.ChangeMask(),
+					filter:         opt.Filter(),
+				})
+				sub.topics[namespace] = struct{}{}
+			}
+
+		case subscriptionID := <-q.unsubscriptionCh:
+			sub, found := q.subscriptions[subscriptionID]
+			if !found {
+				continue
+			}
+
+			for topic := range sub.topics {
+				var updatedFilters []*eventFilter
+				for _, filter := range q.subscriptionsByNS[topic] {
+					if filter.subscriptionID == subscriptionID {
+						continue
+					}
+					updatedFilters = append(updatedFilters, filter)
+				}
+				q.subscriptionsByNS[topic] = updatedFilters
+			}
+
+			delete(q.subscriptions, subscriptionID)
+			delete(q.subscriptionsAll, subscriptionID)
+
+			sub.Kill()
+		}
+	}
+}
+
+func (q *EventQueue) gatherSubscriptions(ch changestream.ChangeEvent) []*subscription {
+	subs := make(map[uint64]*subscription)
+
+	for id := range q.subscriptionsAll {
+		subs[id] = q.subscriptions[id]
+	}
+
+	for _, subOpt := range q.subscriptionsByNS[ch.Namespace()] {
+		// TODO (stickupkid): If the subscription has already been added, do we
+		// want to recall the filter to ensure consistency?
+		if _, ok := subs[subOpt.subscriptionID]; ok {
```

Comment:
Not sure what the TODO is trying to propose. If the event has matched one of the subscription's filters for this topic, then the event should be delivered to that subscription once.

Unless we had negative filters in the future.. or wanted to send the event to the same subscriber once for each passing filter.. then I don't see what else is to be done here.